spark/sql-13/src/main/scala/org/elasticsearch/spark/sql/SchemaUtils.scala

```diff
@@ -273,16 +273,21 @@ private[sql] object SchemaUtils {
     val rowInfo = (new Properties, new Properties)
 
     doDetectInfo(rowInfo, ROOT_LEVEL_NAME, struct)
-    val csv = SettingsUtils.determineSourceFields(settings)
-    // if a projection is applied (filtering or projection) use that instead
-    if (StringUtils.hasText(csv)) {
-      if (settings.getReadMetadata) {
-        rowInfo._1.setProperty(ROOT_LEVEL_NAME, csv + StringUtils.DEFAULT_DELIMITER + settings.getReadMetadataField)
-      }
-      else {
-        rowInfo._1.setProperty(ROOT_LEVEL_NAME, csv)
-      }
-    }
+
+    val requiredFields = settings.getProperty(Utils.DATA_SOURCE_REQUIRED_COLUMNS)
+    val sourceFields = SettingsUtils.determineSourceFields(settings)
+
+    // In case when user selected specific fields (we want to keep the same order in a spark sql row)
+    if (StringUtils.hasText(requiredFields))
+      rowInfo._1.setProperty(ROOT_LEVEL_NAME, requiredFields)
+
+    // In case when we read all fields including metadata
+    else if (StringUtils.hasText(sourceFields) && settings.getReadMetadata)
+      rowInfo._1.setProperty(ROOT_LEVEL_NAME, sourceFields + StringUtils.DEFAULT_DELIMITER + settings.getReadMetadataField)
```

Comment:
This line is a little confusing to me: Is there ever a point in time where any of the `else` statements will be executed? I believe that with the current code `requiredFields` will ALWAYS be set with text, and if it is not, then `sourceFields` will also be empty. I think this could be simplified to just be:

```
val requiredFields = // get required fields
if (StringUtils.hasText(requiredFields)) {
  // set the rowInfo._1 properties
}
```

In the current state it makes me wonder if when we get to this code there is a chance that something prior in the connector wasn't set up entirely, but I don't see where it wouldn't be....